content/posts/2018-08-26-reactive-abstractions.md

```diff
@@ -0,0 +1,294 @@
+            is Request.ShareText -> {
+                val intent = Intent(Intent.ACTION_SEND).apply {
+                    flags = Intent.FLAG_ACTIVITY_NEW_TASK
+                    type = "text/plain"
+                    putExtra(Intent.EXTRA_TEXT, request.text)
+                }
+
+                Intent.createChooser(intent, /* chooser title */ null)
+            }
+        }
+    }
+}
+```
+
+We are controlling the thread under the hood so the consumer does not need to know
+implementation details.
+
+An attentive reader might notice that we are using
+another abstraction — `Application#currentActivity`. It is trivial to implement using
+[`ActivityLifecycleCallbacks`](https://developer.android.com/reference/android/app/Application.ActivityLifecycleCallbacks) and
+`Connectivity`-like approach to listeners and callbacks.
+
+Of course, we can test the behavior.
+
+```kotlin
+class TestLauncher: Launcher {
+    val result = SingleSubject.create<Result>()
+
+    override fun launch(request: Request) = result
+}
+```
+
+```kotlin
+context("launch result is success") {
+
+    beforeEach {
+        launcher.result.onSuccess(Result.Success)
+    }
+
+    it("shows success alert") {
+        verify(view).showSuccessAlert()
+    }
+}
+
+context("launch result is failure") {
+
+    beforeEach {
+        launcher.result.onSuccess(Result.Failure)
+    }
+
+    it("shows failure alert") {
+        verify(view).showFailureAlert()
+    }
+}
+```
+
+## Google Play Services
+
+The concept works well not only with the platform
+but with all third-party information sources — such as various SDK.
+In such cases, it is possible to improve the external API — for example,
+provide proper nullability handling if the SDK is not annotated with `@Nullable` and `@NonNull`.
+
+Let’s say it is necessary to show a warning if Google Play Services
+is not installed.
+
+```kotlin
+disposable += googlePlayServices.available
+    .filter { it == false }
+    .observeOn(mainThread)
+    .subscribe(view::showGooglePlayServicesWarningAlert)
+```
+
+Implementation:
+
+```kotlin
+interface GooglePlayServices {
+
+    val available: Single<Boolean>
+
+    class PackagedGooglePlayServices(
+        context: AndroidContext,
+        ioScheduler: Scheduler
+    ) : GooglePlayServices {
+
+        override val available = Single
+            .fromCallable { GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context) }
+            .map { it == ConnectionResult.SUCCESS }
+            .subscribeOn(ioScheduler)
+    }
+}
+```
+
+The work is done on IO thread since retrieving the information can take an undefined amount of time.
+Since the RxJava is used all multi-threading mumbo-jumbo is handled automagically.
+
+And of course, it is possible to test the behavior without mocking static
+`GoogleApiAvailability` `getInstance()` method.
+
+# Retrospective
+
+The beautiful thing about abstractions is that they are universal and can be applied everywhere.
```

Comment:
yet it's unclear if such platform abstractions should be open sourced and shared (just a thought)